### PR TITLE
support imx6ull dcp crypto driver

### PR DESCRIPTION
**Pull request by mike-scott**
- bring in the upstream PR for supporting DCP from here (squashed): https://github.com/OP-TEE/optee_os/pull/4094
- add @ldts patch to set DCP exclusive access for OP-TEE (disable for kernel)

### CODE REVIEW

**Comment by mike-scott** [comment]
rebased

**Comment by ldts** [comment]
how did you test? also just wondering, is this HUK only readable once the board is closed and in the meantime it only generates a development HUK (like the CAAM)?

**Comment by mike-scott** [comment]
I'm testing now.  This is very WIP.

**Comment by ricardosalveti** [comment]
Nice that the extra changes are well isolated to imx6ul, so +1 (once you test and verify that it works on the final target).

**Comment by ldts** [comment]
can you check /sys/firmware/devicetree/base/soc@0/aips-bus@2200000/crypto@2280000 ? just the secure-status and status properties to make sure the reservation happened

**Comment by ldts** [comment]
not that this reservation will be broken on latest kernels as discussed here https://github.com/OP-TEE/optee_os/pull/4277
hopefully the kernel will not try to enable it by default

**Comment by mike-scott** [comment]
Added 1 commit to fix the DCP HUK support.

**Comment by ricardosalveti** [comment]
> not that this reservation will be broken on latest kernels as discussed here [OP-TEE#4277](https://github.com/OP-TEE/optee_os/pull/4277)
> hopefully the kernel will not try to enable it by default

We can have 2 lines here right? One for the current path and another for what is currently used by linux upstream, that way this won't bite us when updating to a newer kernel.

**Comment by ricardosalveti** [comment]
> Added 1 commit to fix the DCP HUK support.

This is the set that got tested by you right? Or do you have any other extra/pending changes?

**Comment by mike-scott** [comment]
> > Added 1 commit to fix the DCP HUK support.
> 
> This is the set that got tested by you right? Or do you have any other extra/pending changes?

Yes, this is the patch that I tested and submitted back upstream.

**Comment by ricardosalveti** [comment]
> > not that this reservation will be broken on latest kernels as discussed here [OP-TEE#4277](https://github.com/OP-TEE/optee_os/pull/4277)
> > hopefully the kernel will not try to enable it by default
> 
> We can have 2 lines here right? One for the current path and another for what is currently used by linux upstream, that way this won't bite us when updating to a newer kernel.

Pinging @ldts directly so we don't forget. Looks like we can merge this then once we sort out the dtb path (just to avoid having this issue again later on once we bump to 5.10, which should probably be released by NXP soon).

**Review by Tim-Anderson — APPROVED** [review]
LGTM

File: core/arch/arm/plat-imx/main.c [hunk 1]
```diff
@@ -114,6 +114,25 @@ void console_init(void)
...
 }
 
+/* reserve the following drivers for exclusive OP-TEE access */
+const char *main_get_optee_exclusive_node_name(unsigned int i)
+{
+	static const char * const exclusive_drivers[] = {
+#if defined CFG_IMX_DCP
+/* list platforms */
+#if defined(PLATFORM_FLAVOR_mx6ullevk)
+		"/soc/aips-bus@2200000/crypto@2280000",
```
Comment (ricardosalveti) [inline core/arch/arm/plat-imx/main.c:124]: Can you also cover "/soc/bus@2200000/crypto@2280000" here so we can make it compatible with latest kernel at the same time?

**Comment by mike-scott** [comment]
Updated with @ldts latest version of exclusive access patch.

Diff from previous version:
```
diff --git a/core/arch/arm/plat-imx/main.c b/core/arch/arm/plat-imx/main.c
index 56a30006..2bf36317 100644
--- a/core/arch/arm/plat-imx/main.c
+++ b/core/arch/arm/plat-imx/main.c
@@ -122,6 +122,7 @@ const char *main_get_optee_exclusive_node_name(unsigned int i)
 /* list platforms */
 #if defined(PLATFORM_FLAVOR_mx6ullevk)
                "/soc/aips-bus@2200000/crypto@2280000",
+               "/soc/bus@2200000/crypto@2280000",
 #endif
 #endif
                NULL,
```

**Review by ricardosalveti — APPROVED** [review]
LGTM

**Pull request merged**